splunk-cloudwatch-logs-processor/index.js

```diff
@@ -185,16 +185,58 @@ const CloudWatchToSplunk = (parsed, context, logger, sourcetype, callback) => {
             - Set or remove metadata properties as needed. For descripion of each property, refer to:
             http://docs.splunk.com/Documentation/Splunk/latest/RESTREF/RESTinput#services.2Fcollector */
 
-            const log = {
-                message: item.message,
-                metadata: {
-                    time: item.timestamp ? new Date(item.timestamp).getTime() / 1000 : Date.now(),
-                    host: parsed.logGroup,
-                    source: parsed.logStream,
-                    sourcetype: sourcetype,
-                    //index: 'main',
-                },
-             };
+            let log;
+
+            try {
+                log = JSON.parse(item.message);
+            }
+
+            // Ignore SyntaxError exception if message not in JSON format
+            catch (e) { }
+
+            // Check if message is already in JSON format
+            if (typeof log === "object" && message in log) {
+                // Add metadata fields if not present
+                if (metadata in log) {
```

Comment:
I have some concerns about modifying the metadata if it exists. We will not know exactly what to expect in the metadata. @JonRoma and @kwessel is it feasible to encapsulate the JSON object in the message field?

Reply:
I'm thinking the whole thing could be reduced to:

```javascript
            let message;

            try {
                message = JSON.parse(item.message);
            } catch (e) {
                message = item.message;
            }
            const log = {
                message: message,
                metadata: {
                    time: item.timestamp ? new Date(item.timestamp).getTime() / 1000 : Date.now(),
                    host: parsed.logGroup,
                    source: parsed.logStream,
                    sourcetype: sourcetype,
                    //index: 'main',
                },
             };
```

But maybe this is not useful, but it has the advantage that the JSON object logged is never modified and the metadata has a clear meaning.

Reply:
> I have some concerns about modifying the metadata if it exists. We will not know exactly what to expect in the metadata. @JonRoma and @kwessel is it feasible to encapsulate the JSON object in the message field?

David, this tells me that I did a terrible job of making the point of this PR unclear.

Consider the case where we have two different containers logging in a service, both to streams in the same log group. They've got different formats because they're different applications. Thus, we need to pass in a different source type for logs from each container.

my PR would allow for JSON in the message field, but your existing Lambda code already does that because Splunk is smart enough to see that a JSON object encapsulated in a string in the message field is still JSON. But your code doesn't allow us to set different source types for different events. Thus the desire to include the metadata object in the JSON coming out of the container so we can set any of the four metadata fields (one of which is source type) on a per-event basis.

As it is now, the Splunk folks have had to do ingest-side transforms for logs coming out of Shibboleth, and Shibboleth has to identify what application and log file the event is from by prepending a bunch of semicolon-separated values onto the front of the event. With my PR, Shibboleth can spit out this data in JSON format (in the metadata object), and the Splunk folks won't have to do anything. Ryan has requested that we start including the source type in the log event so that they don't have to keep doing this very not scalible per-service ingest-time transform coding. This puts the defining of meta fields on a per-event basis into the hands of the service owner.

But the function will still fall through to defaults if the metadata isn't completely present or, if the event isn't JSON, will revert to what you were already doing in your code.

Given that explanation, if you're still uncomfortable with modifying the metadata object, I'd be happy to explore other solutions to this problem with you. But hopefully, that clears things up.

Keith